src/main/java/codeu/model/store/basic/MessageStore.java

```diff
@@ -56,37 +59,115 @@ public static MessageStore getTestInstance(PersistentStorageAgent persistentStor
    */
   private PersistentStorageAgent persistentStorageAgent;
 
-  /** The in-memory list of Messages. */
-  private List<Message> messages;
+  /** The in-memory map of Messages. */
+  private Map<UUID, Message> messages;
+
+  // These map from some conversation/user ID to a list containing the IDs of their messages
+  private Map<UUID, List<UUID>> conversationIdToMessageIdList;
+  private Map<UUID, List<UUID>> authorIdToMessageIdList;
 
   /** This class is a singleton, so its constructor is private. Call getInstance() instead. */
   private MessageStore(PersistentStorageAgent persistentStorageAgent) {
     this.persistentStorageAgent = persistentStorageAgent;
-    messages = new ArrayList<>();
+    messages = new HashMap<>();
+
+    conversationIdToMessageIdList = new HashMap<>();
+    authorIdToMessageIdList = new HashMap<>();
   }
 
-  /** Add a new message to the current set of messages known to the application. */
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This writes the message to the persistent storage.
+   */
   public void addMessage(Message message) {
-    messages.add(message);
+    addMessageWithoutPersistentStorage(message);
+
     persistentStorageAgent.writeThrough(message);
   }
 
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This does NOT write the message to the persistent storage.
+   */
+  public void addMessageWithoutPersistentStorage(Message message) {
+    UUID id = message.getId();
+    UUID conversationId = message.getConversationId();
+    UUID authorId = message.getAuthorId();
+
+    messages.put(id, message);
+
+    // Add to its conversation's list
```

Comment:
It might make sense to factor out the logic below into a static private method so it isn't repeated.

Something like addToListInMap(mapName, mapKey, listValue)

This has the benefit of being more readable for someone scanning the code, they don't have to parse the if statements in their head. Your code would then basically read like your comments which is ideal!

Reply:
Done.

Reply:
I would say the comments are redundant at this point as they simply repeat what the code says. This is a good thing! In an ideal world we wouldn't need comments because all code would be self-explanatory.